Positional sounds strength adjustment slider implementation

mk56-spn opened the pull request:
As the title says I have replaced (technically, read last parragraph )  the toggle for positional hitsounds with an adjustable slider.

I should note that the intensity is mutliplied by 2 to make the scaling stronger as it goes up , since at 1x  it felt weak.

Furthermore I have not yet deleted the old binding logic for the toggle incase for some reason it is deemed neccesary, upon confirmation that it isnt i will axe it. This is a small QoL change hence the lack of a discussion over it.

mk56-spn committed:
Initial implementation of adjustable positional hitobject audio strength

mk56-spn committed:
Merge branch 'ppy:master' into positional-sounds-strength-adjustment

mk56-spn committed:
Binding logic fix, nullification of unnecessary path

mk56-spn committed:
Merge remote-tracking branch 'origin/positional-sounds-strength-adjustment' into positional-sounds-strength-adjustment

mk56-spn committed:
Merge remote-tracking branch 'origin/positional-sounds-strength-adjustment' into positional-sounds-strength-adjustment

mk56-spn committed:
cleanup

mk56-spn committed:
Merge remote-tracking branch 'origin/positional-sounds-strength-adjustment' into positional-sounds-strength-adjustment

mk56-spn commented:
Here is how it looks in game :

PercyDan54 commented on osu.Game/Overlays/Settings/Sections/Gameplay/AudioSettings.cs:
```diff
@@ -2,27 +2,51 @@
+
         [BackgroundDependencyLoader]
-        private void load(OsuConfigManager config)
+        private void load(OsuConfigManager config,OsuConfigManager osuConfig)
```
Why duplicate two `OsuConfigManager`

mk56-spn replied:
First of all it wouldn't kill you to be less rude. And the reason I duplicated it was because I was unaware of how to direct osuconfigmanager to config and osuconfig at once.

bdach replied:
the name of the injected variable doesn't matter. it's fine to just use `config` everywhere, there is no difference between the two

PercyDan54 commented on osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs:
```diff
@@ -532,9 +535,9 @@ protected virtual void ApplySkin(ISkinSource skin, bool allowFallback)
+            float balance_adjust_amount = positionalHitsoundsLevel.Value*2;
 
-            return balance_adjust_amount * (userPositionalHitSounds.Value ? position - 0.5f : 0);
+            return balance_adjust_amount * (true ? position - 0.5f : 0);
```
Wtf?

mk56-spn replied:
That was simply a placeholder whilst I wait on bdach or peppy to say it's ok to delete the old binding logic . Nothing else

PercyDan54 replied:
This won't pass code quality checks, and there's a missing whitespace at L538

mk56-spn replied:
> This won't pass code quality checks, and there's a missing whitespace at L538

Are you at all reading what I'm saying?. I'm telling you it's a placeholder. I'll fix the whitespace.

bdach replied:
the `true` should just be removed, correct. suffices to use the position unconditionally.

PercyDan54 reviewed (commented):
Have you ever read your code again before PR?

bdach commented on osu.Game/Localisation/AudioSettingsStrings.cs:
```diff
@@ -32,6 +32,11 @@ public static class AudioSettingsStrings
         /// <summary>
         /// "Master"
         /// </summary>
+        public static LocalisableString PositionalLevel => new TranslatableString(getKey(@"positional_hitsound_audio_level"), @"Positional hitsound audio level.");
```
not sure how this happened but the xmldocs here are completely wrong. did you add this entry manually? it should not be done manually, there is tooling that is supposed to do it automatically for you.

additionally, no other string in this file is terminated with a full stop:

```suggestion
        public static LocalisableString PositionalLevel => new TranslatableString(getKey(@"positional_hitsound_audio_level"), @"Positional hitsound audio level");
```

bdach commented on osu.Game/Configuration/OsuConfigManager.cs:
```diff
@@ -97,6 +97,7 @@ protected override void InitialiseDefaults()
             SetDefault(OsuSetting.MenuParallax, true);
 
             // Gameplay
+            SetDefault(OsuSetting.PositionalHitsoundsLevel, 0.8f, 0, 1);
```
the default here changes behaviour. was that intentional? as far as i can tell, to preserve the previous balance it should be `0.2f` (because the value is doubled at usage site)

bdach reviewed (changes requested):
not sure whether this should be a thing that's configurable, but here's a minimum review for this to even be viable to consider merging.

mk56-spn committed:
Fixed the problems that were brought up and deleted the old bind logic

mk56-spn committed:
Fix

mk56-spn committed:
Fix

mk56-spn committed:
Merge remote-tracking branch 'origin/positional-sounds-strength-adjustment' into positional-sounds-strength-adjustment

peppy commented:
Thanks for the attempted contribution, but this doesn't align with our quality expectations, and also adds functionality without previous discussion.

There's still many issue with your code quality and formatting (please read the contribution guides), but looking past that:
- You've removed an existing setting with no migration. Users which disabled positional audio will have their setting lost.
- There was no discussion on the UX of this. You've taken a simple toggle and made it into something which is more complex, and something which was potentially never in demand in the first place. Generally we ask you make discussions first.
- The text and setting in general don't read well.

peppy reviewed (changes requested):
Not sure whether to close or leave for further perseverance. Really dislike people making PRs with changes like this without previous discussion.

mk56-spn commented:
Understandable. The only reason I didn't make a discussion is because you said it was fine for the other tweak I made. I guess I failed to consider the difference in scope. 

I've attempted to get the code formatting working to no avail. As for the saving the users previous selections I didn't even think of that. Not really an excuse just yeah.

mk56-spn commented:
> Understandable. The only reason I didn't make a discussion is because you said it was fine for the other tweak I made. I guess I failed to consider the difference in scope. 
> 
> I've attempted to get the code formatting working to no avail. As for the saving the users previous selections I didn't even think of that. Not really an excuse just yeah. 

I do atleast hope to atleast make it clear that I didn't not open a discussion out of laziness. But  because I thought it would be less work for you guys.

mk56-spn committed:
Merge branch 'ppy:master' into positional-sounds-strength-adjustment

mk56-spn commented:
> Not sure whether to close or leave for further perseverance. Really dislike people making PRs with changes like this without previous discussion.

gonna give this one final lookover

migration part i have no idea how to do but i know what needs to be done

as for the UX/text aspect would you mind elaborating on what rubbed you the wrong way? the location in settings? the values on the bar? idk cant think of much here

peppy commented:
For the migration, you might be able to get some ideas from existing settings migrations (https://github.com/ppy/osu/blob/1aac978a7b5a2c866de797ea53a9f7a6aeac0b70/osu.Game/Configuration/OsuConfigManager.cs#L161).

For the text, it shouldn't have a `.` at the end and should be called something like "Hitsound stereo separation" or something

mk56-spn committed:
Code refactor and name changes

cleaned code up with Jetbrains i hope it suffices

mk56-spn commented:
migrating is harder than I expected and i havent been able to do it . but all the other issues with code quality should be adressed

peppy committed:
Merge branch 'master' into positional-sounds-strength-adjustment

peppy committed:
Remove outdated settings migration

peppy committed:
Add migration of positional hitsounds setting to new level based setting

peppy committed:
Remove unnecessary code and fix double nesting causing filtering to not work

peppy commented:
I've added the required migrations and fixed up a few remaining code issues here. Looks good to go now 👍 .

mk56-spn commented:
> I've added the required migrations and fixed up a few remaining code issues here. Looks good to go now 👍 .

I'll look through the migrations so I can do them in the future. 

Btw was the post jetbrains clean up code noticeably better? Just want to know for reference .

peppy commented:
Take a look at https://github.com/ppy/osu/pull/15844/commits/6356180b6abd5a11ab32508ad3dcd2984c561aee for the issues I fixed. Notably, the bindable didn't need to be stored to a local field, and you nested the new setting one fill flow level too deep, which meant it wasn't properly showing up when searching/filtering in settings.
